llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -0,0 +1,433 @@
+      EntryPointGroups.push_back({EPG.first, std::move(EPG.second)});
+  } else {
+    // No entry points met, record this.
+    EntryPointGroups.push_back({GLOBAL_SCOPE_NAME, {}});
+  }
+
+  return EntryPointGroups;
+}
+
+// For device global variables with the 'device_image_scope' property,
+// the function checks that there are no usages of a single device global
+// variable from kernels grouped to different modules.
+void checkImageScopedDeviceGlobals(const Module &M,
+                                   const EntryPointGroupVec &Groups) {
+  // Early exit if there is only one group
+  if (Groups.size() < 2)
+    return;
+
+  // Reverse the EntryPointGroupMap to get a map of entry point -> module's name
+  unsigned EntryPointNumber = 0;
+  for (const auto &Group : Groups)
+    EntryPointNumber += static_cast<unsigned>(Group.Functions.size());
+  DenseMap<const Function *, StringRef> EntryPointModules(EntryPointNumber);
+  for (const auto &Group : Groups)
+    for (const auto *F : Group.Functions)
+      EntryPointModules.insert({F, Group.GroupId});
+
+  // Processing device global variables with the "device_image_scope" property
+  for (auto &GV : M.globals()) {
+    if (!isDeviceGlobalVariable(GV) || !hasDeviceImageScopeProperty(GV))
+      continue;
+
+    Optional<StringRef> VarEntryPointModule{};
+    auto CheckEntryPointModule = [&VarEntryPointModule, &EntryPointModules,
+                                  &GV](const auto *F) {
+      auto EntryPointModulesIt = EntryPointModules.find(F);
+      assert(EntryPointModulesIt != EntryPointModules.end() &&
+             "There is no group for an entry point");
+      if (!VarEntryPointModule.hasValue()) {
+        VarEntryPointModule = EntryPointModulesIt->second;
+        return;
+      }
+      if (EntryPointModulesIt->second != *VarEntryPointModule) {
+        error("device_global variable '" + Twine(GV.getName()) +
+              "' with property \"device_image_scope\" is used in more "
+              "than one device image.");
+      }
+    };
+
+    SmallSetVector<const User *, 32> Workqueue;
+    for (auto *U : GV.users())
+      Workqueue.insert(U);
+
+    while (!Workqueue.empty()) {
+      const User *U = Workqueue.back();
+      Workqueue.pop_back();
+      if (auto *I = dyn_cast<const Instruction>(U)) {
+        auto *F = I->getFunction();
+        Workqueue.insert(F);
+        continue;
+      }
+      if (auto *F = dyn_cast<const Function>(U)) {
+        if (EntryPointModules.count(F))
+          CheckEntryPointModule(F);
+      }
+      for (auto *UU : U->users())
+        Workqueue.insert(UU);
+    }
+  }
+}
+
+void collectFunctionsToExtract(SetVector<const GlobalValue *> &GVs,
+                               const EntryPointGroup &ModuleEntryPoints) {
+  for (const auto *F : ModuleEntryPoints.Functions)
+    GVs.insert(F);
+
+  decltype(GVs.size()) I = 0;
+  while (I != GVs.size()) {
+    const auto *F = cast<Function>(GVs[I++]);
+    for (const auto &I : instructions(F))
```

Comment:
This `I` shadows the first `I` what is an index. Maybe we should rename the first `I` in `Var` or `Idx`, or something like this?